### PR TITLE
Add scripts to support test execution with sanitizers enabled.

### PR DESCRIPTION
**Pull request by oschaaf**
- Refactors run-ci.sh a little bit to make it easier to use in dev flows
- Add new scripts to support executing with asan/tsan/msan enabled

Signed-off-by: Otto van der Schaaf <ovanders@redhat.com>

### CODE REVIEW

File: maistra/run-ci.sh [hunk 1]
```diff
@@ -1,36 +1,7 @@
...
-if [ "${ARCH}" = "ppc64le" ]; then
-  ARCH="ppc"
-fi
-export ARCH
-
-export BUILD_SCM_REVISION="Maistra PR #${PULL_NUMBER:-undefined}"
-export BUILD_SCM_STATUS="SHA=${PULL_PULL_SHA:-undefined}"
-
-COMMON_FLAGS="\
-    --config=clang \
```
Comment (oschaaf) [inline maistra/run-ci.sh:19]: Note: we drop `--config=clang` in this change; bit I think that's fine, as we seem to default to it anyway.

File: maistra/test-with-asan.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+FLAGS="${COMMON_FLAGS} \
+  --config=clang-${SANITIZER} \
+  --build_tests_only \
+  --test_env=ENVOY_IP_TEST_VERSIONS=v4only \
+  --test_output=all \
+  ${BAZEL_TESTS} \
+"
+
+echo "Build tests with ${SANITIZER:asan} enabled."
```
Comment (oschaaf) [inline maistra/test-with-asan.sh]: Note to self: add a comment to explain why we build & test in separate steps here

**Comment by oschaaf** [comment]
Hmm the CI logs seem to have a lot of warnings related to writes to the build cache [1], but I'm not sure how these relate to this change.. 

[1] 
```
WARNING: Writing to Remote Cache:
com.google.devtools.build.lib.remote.BulkTransferException: 4 errors during bulk transfer
	at com.google.devtools.build.lib.remote.RemoteCache.waitForBulkTransfer(RemoteCache.java:234)
	at com.google.devtools.build.lib.remote.RemoteCache.uploadOutputs(RemoteCache.java:208)
	at com.google.devtools.build.lib.remote.RemoteCache.upload(RemoteCache.java:142)
	at com.google.devtools.build.lib.remote.RemoteCache.upload(RemoteCache.java:160)
	at com.google.devtools.build.lib.remote.RemoteSpawnCache$1.store(RemoteSpawnCache.java:288)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:145)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:102)
	at com.google.devtools.build.lib.actions.SpawnStrategy.beginExecution(SpawnStrategy.java:47)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.beginExecution(SpawnStrategyResolver.java:65)
	at com.google.devtools.build.lib.rules.cpp.CppCompileAction.beginExecution(CppCompileAction.java:1452)
	at com.google.devtools.build.lib.actions.Action.execute(Action.java:127)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$5.execute(SkyframeActionExecutor.java:855)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:1016)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:975)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:129)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:81)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:472)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:834)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:307)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:477)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
	Suppressed: com.google.devtools.build.lib.remote.http.HttpException: 500 Internal Server Error
...
```

File: maistra/run-ci.sh (hunk 1 above)
Comment (jwendell) [reply maistra/run-ci.sh:19]: What do you mean? The default compiler is clang, but this config doesn't seem to be picked up automatically

File: maistra/common.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+export ARCH
+
+export BUILD_SCM_REVISION="Maistra PR #${PULL_NUMBER:-undefined}"
+export BUILD_SCM_STATUS="SHA=${PULL_PULL_SHA:-undefined}"
+export CI_CONFIG=${CI_CONFIG:-1}
+
+COMMON_FLAGS="\
+    --config=${ARCH} \
+"
+if [ "${CI_CONFIG}" -eq "1" ]; then
```
Comment (jwendell) [inline maistra/common.sh]: why not use the `CI` variable? It's set to `prow` in our [images](https://github.com/maistra/test-infra/blob/main/docker/maistra-builder_2.2.Dockerfile#L55). I think that just checking if it's not empty is enough.

File: maistra/test-with-asan.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+SANITIZER=${SANITIZER:-asan}
+BAZEL_TESTS=${1:-//test/...}
+
+FLAGS="${COMMON_FLAGS} \
+  --config=clang-${SANITIZER} \
+  --build_tests_only \
+  --test_env=ENVOY_IP_TEST_VERSIONS=v4only \
+  --test_output=all \
+  ${BAZEL_TESTS} \
```
Comment (jwendell) [inline maistra/test-with-asan.sh:24]: could we also move these three common parameters to `maistra/bazelrc` under a `test` line? (then we need to remove them from `run-ci.sh` as well.

File: maistra/common.sh (hunk 3 above)
Comment (oschaaf) [reply maistra/common.sh]: 👍 will change this to use the `CI`variable.

File: maistra/test-with-asan.sh (hunk 4 above)
Comment (oschaaf) [reply maistra/test-with-asan.sh:24]: We could.  Looking at this once more, I think these actually should be CI-specific as well. (I don't think there's anything against executing ipv6 tests on a (dev-) system that supports it).

File: maistra/test-with-asan.sh [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+# An argument can be passed to this script which when specified will replace
+# the "//test/..." part we pass to bazel. This allows one to optionally
+# specify which tests to run.
+
+# Furthermore, while this defaults to asan, one can set the SANITIZER env var
+# to specify other sanitizers, as support in our bazel setup. 
+# As of writing this, known valid values are "tsan" and "msan".
+
+# Lastly, when running this in development flows, one might want to set BAZEL_CONFIG=
```
Comment (jwendell) [inline maistra/test-with-asan.sh]: `s/BAZEL_CONFIG/CI/` ?

File: maistra/test-with-asan.sh (hunk 2 above)
Comment (jwendell) [reply maistra/test-with-asan.sh]: I'm curious

File: maistra/common.sh [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+export CC=clang CXX=clang++
+
+ARCH=$(uname -p)
+if [ "${ARCH}" = "ppc64le" ]; then
+  ARCH="ppc"
+fi
+export ARCH
+
+export BUILD_SCM_REVISION="Maistra PR #${PULL_NUMBER:-undefined}"
+export BUILD_SCM_STATUS="SHA=${PULL_PULL_SHA:-undefined}"
```
Comment (jwendell) [inline maistra/common.sh]: I think these two only make sense inside CI

File: maistra/run-ci.sh (hunk 1 above)
Comment (oschaaf) [reply maistra/run-ci.sh:19]: The build succeeds, so I am not going to add `--config=clang` back as per offline discussion

File: maistra/test-with-asan.sh (hunk 2 above)
Comment (oschaaf) [reply maistra/test-with-asan.sh]: Commented about it in 63e6ee7051427382c107bc6e7621530c579d2eed

File: maistra/test-with-asan.sh (hunk 5 above)
Comment (oschaaf) [reply maistra/test-with-asan.sh]: Fixed in 63e6ee7051427382c107bc6e7621530c579d2eed

File: maistra/common.sh (hunk 6 above)
Comment (oschaaf) [reply maistra/common.sh]: Good catch -- moved these back into `maistra/run-ci.sh`

**Review by jwendell — APPROVED** [review]

**Pull request merged**